### PR TITLE
Adding support for Elasticsearch 6.

### PR DESCRIPTION
**Pull request by dobooth**
## This PR is a:

- Content update

## Summary

When this pull request is merged, it will...

Update docs to add support for Elasticsearch 6.

whatsnew
Added information about [Elasticsearch 6](https://devdocs.magento.com/guides/v2.3/config-guide/elasticsearch/es-overview.html#es-spt-versions) support.
Added Elasticsearch 6 information to [Configure Magento to use Elasticsearch](https://devdocs.magento.com/guides/v2.3/config-guide/elasticsearch/configure-magento.html)
Added [Change Elasticsearch Client](https://devdocs.magento.com/guides/v2.3/config-guide/elasticsearch/es-downgrade.html) topic.
Updated the [Technologies Magento can use](https://devdocs.magento.com/guides/v2.3/install-gde/system-requirements-tech.html#technologies-magento-can-use) topic with new Elasticsearch 6 support.

### CODE REVIEW

File: guides/v2.3/config-guide/elasticsearch/es-downgrade.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+Both ES 2.x and 5.x are [End of Life][].
+Running ES 2.x is strongly discouraged.
+
+## Change the Elasticsearch Client version
+
+If you need to work with Elasticsearch 5.x, run the following command:
+
+```bash
+composer require "elastic/elasticsearch:~5.1"
```
Comment (keharper) [inline guides/v2.3/config-guide/elasticsearch/es-downgrade.md]: We previously claimed support for 5.2. Check if this version can be bumped up to 5.2

File: _includes/config/es-elasticsearch-magento.md [hunk 2]
```diff
@@ -1,102 +1,96 @@
...
-1.  In the Admin, click <strong>System</strong> > <strong>Cache Management</strong>.
-2.  Select the checkbox next to <strong>Page Cache</strong>.
-3.  From the <strong>Actions</strong> list in the upper right, click <strong>Refresh</strong>.<br />
-		The following figure shows an example.<br />
-		![]({{ site.baseurl }}/common/images/solr_refresh-cache.png){:width="600px"}
+1. In the Admin, click **System** > **Cache Management**.
+1. Select the checkbox next to **Page Cache**.
+1. From the **Actions** list in the upper right, click **Refresh**.
+   The following figure shows an example.
+   ![]({{ site.baseurl }}/common/images/solr_refresh-cache.png){:width="600px"}
```
Comment (keharper) [inline _includes/config/es-elasticsearch-magento.md]: Have an architect review this graphic. We no longer support solr

File: guides/v2.2/install-gde/system-requirements-tech.md [hunk 3]
```diff
@@ -82,53 +83,56 @@ For more information, see [Required PHP settings]({{ page.baseurl }}/install-gde
...
+#### {{site.data.var.ee}} only
 
-*	Elasticsearch
+* Elasticsearch
 
     {{site.data.var.ee}} version 2.2.x supports the following Elasticsearch versions:
 
-    *	Elasticsearch [5.x](https://www.elastic.co/downloads/past-releases/elasticsearch-5-2-2){:target="_blank"}
-    *	Elasticsearch [2.x](https://www.elastic.co/downloads/past-releases/elasticsearch-2-4-5){:target="_blank"}
+  * Elasticsearch [6.x](https://www.elastic.co/downloads/past-releases/elasticsearch-6-6-1){:target="_blank"}
```
[the remark this answers is not in the document]
Comment (dobooth) [reply guides/v2.2/install-gde/system-requirements-tech.md]: Done.

File: guides/v2.3/config-guide/elasticsearch/es-downgrade.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
+---
+group: configuration-guide
+title: Change Elasticsearch Client
```
[the remark this answers is not in the document]
Comment (dobooth) [reply guides/v2.3/config-guide/elasticsearch/es-downgrade.md]: Done.

File: guides/v2.3/config-guide/elasticsearch/es-downgrade.md (hunk 1 above)
Comment (dobooth) [reply guides/v2.3/config-guide/elasticsearch/es-downgrade.md]: Good catch.

File: _includes/config/es-elasticsearch-magento.md (hunk 2 above)
Comment (dobooth) [reply _includes/config/es-elasticsearch-magento.md]: Not solr related..just the file name. Admin looks the same.

File: guides/v2.3/config-guide/elasticsearch/es-downgrade.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+functional_areas:
+  - Configuration
+  - Search
+  - System
+  - Setup
+---
+
+The Magento 2.3.1 update adds support for Elasticsearch (ES) 6.x.
+Magento still provides connectivity for ES 2.x and 5.x but this will need to be specifically enabled.
```
[the remark this answers is not in the document]
Comment (dobooth) [reply guides/v2.3/config-guide/elasticsearch/es-downgrade.md]: Done.

File: _includes/config/es-elasticsearch-magento.md [hunk 6]
```diff
@@ -1,102 +1,96 @@
...
+   (The **Elasticsearch 5.0+/6.0+** options are not available for Magento 2.1.)
+
+![]({{ site.baseurl }}/common/images/elastic_choose-in-admin-es6.png){:width="650px"}
+
+The following table discusses only the configuration options required to test the connection with Magento.
+Unless you changed Elasticsearch server settings, the defaults should work. Skip to the next step.
+
+|Option|Description|
+|--- |--- |
+|Elasticsearch Server Hostname|Enter the fully qualified hostname or IP address of the machine running Elasticsearch. {{site.data.var.ece}}: Get this value from your integration system.|
```
[the remark this answers is not in the document]
Comment (dobooth) [reply _includes/config/es-elasticsearch-magento.md]: Done.

File: _includes/config/es-elasticsearch-magento.md [hunk 7]
```diff
@@ -1,102 +1,96 @@
...
+1. Click **Test Connection**.
+
+You will see either:
+
+<img src="{{ site.baseurl }}/common/images/elastic_test-success.png">
+
+Magento successfully connected to the Elasticsearch server.
+Continue with:
+
+* [Configure Apache and Elasticsearch]({{ site.gdeurl21 }}config-guide/elasticsearch/es-config-apache.html)
```
[the remark this answers is not in the document]
Comment (dobooth) [reply _includes/config/es-elasticsearch-magento.md]: Changed to baseurl.

File: guides/v2.1/config-guide/elasticsearch/es-downgrade.md [hunk 8]
```diff
@@ -0,0 +1,17 @@
+---
+group: configuration-guide
+title: Change Elasticsearch Client
```
[the remark this answers is not in the document]
Comment (dobooth) [reply guides/v2.1/config-guide/elasticsearch/es-downgrade.md]: Yup.

File: guides/v2.1/config-guide/elasticsearch/es-downgrade.md [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+functional_areas:
+  - Configuration
+  - Search
+  - System
+  - Setup
+---
+
+The Magento 2.1 only supports Elasticsearch 2.0, which has reached [End of Life][].
```
[the remark this answers is not in the document]
Comment (dobooth) [reply guides/v2.1/config-guide/elasticsearch/es-downgrade.md]: Got it.

File: guides/v2.2/config-guide/elasticsearch/es-downgrade.md [hunk 10]
```diff
@@ -0,0 +1,44 @@
+---
+group: configuration-guide
+title: Change Elasticsearch Client
```
[the remark this answers is not in the document]
Comment (dobooth) [reply guides/v2.2/config-guide/elasticsearch/es-downgrade.md]: Done.

File: guides/v2.2/config-guide/elasticsearch/es-downgrade.md [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+functional_areas:
+  - Configuration
+  - Search
+  - System
+  - Setup
+---
+
+The Magento 2.2.8 update adds support for Elasticsearch (ES) 6.x.
```
[the remark this answers is not in the document]
Comment (dobooth) [reply guides/v2.2/config-guide/elasticsearch/es-downgrade.md]: Got it.

File: guides/v2.2/config-guide/elasticsearch/es-downgrade.md [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+Magento still provides connectivity for ES 2.x and 5.x but this will need to be specifically enabled.
```
[the remark this answers is not in the document]
Comment (dobooth) [reply guides/v2.2/config-guide/elasticsearch/es-downgrade.md]: Good.

File: guides/v2.2/config-guide/elasticsearch/es-overview.md [hunk 13]
```diff
@@ -9,198 +9,152 @@ functional_areas:
...
-
-    ```
-		[elasticsearch-2.x]
-		name=Elasticsearch repository for 2.x packages
-		baseurl=http://packages.elastic.co/elasticsearch/2.x/centos
-		gpgcheck=1
-		gpgkey=http://packages.elastic.co/GPG-KEY-elasticsearch
-		enabled=1
-    ```
+## Upgrade from Elasticsearch 2.x/5x to 6.x {#es-upgrade6}
```
[the remark this answers is not in the document]
Comment (dobooth) [reply guides/v2.2/config-guide/elasticsearch/es-overview.md]: Fixed.

File: guides/v2.2/install-gde/system-requirements-tech.md [hunk 14]
```diff
@@ -82,53 +83,56 @@ For more information, see [Required PHP settings]({{ page.baseurl }}/install-gde
...
 
 Mail Transfer Agent (MTA) or an SMTP server
 
-### Magento can use the following technologies:
+### Magento can use the following technologies
 
-*	[Redis]({{ page.baseurl }}/config-guide/redis/config-redis.html) version 3.2 (compatible with 2.4+ ) for page caching and session storage
-*	[Varnish]({{ page.baseurl }}/config-guide/varnish/config-varnish.html) version 4.x or 5.0
-*	[memcached]({{ page.baseurl }}/config-guide/memcache/memcache.html) latest stable version for session storage with either `memcache` or `memcached` PHP extensions (latest stable version)
+* [Redis]({{ page.baseurl }}/config-guide/redis/config-redis.html) version 3.2 (compatible with 2.4+ ) for page caching and session storage
```
[the remark this answers is not in the document]
Comment (dobooth) [reply guides/v2.2/install-gde/system-requirements-tech.md]: Updated.

File: guides/v2.2/install-gde/system-requirements-tech.md [hunk 15]
```diff
@@ -48,25 +49,25 @@ Magento only uses MySQL features compatible with MariaDB. MariaDB may not be com
...
-*   [PDO/MySQL](http://php.net/manual/en/ref.pdo-mysql.php){:target="_blank"}
-*   [SimpleXML](http://php.net/manual/en/book.simplexml.php){:target="_blank"}
-*   [soap](http://php.net/manual/en/book.soap.php){:target="_blank"}
-*   [spl](http://php.net/manual/en/book.spl.php){:target="_blank"}
-*   [libxml](http://php.net/manual/en/book.libxml.php){:target="_blank"}
-*   [xsl](http://php.net/manual/en/book.xsl.php){:target="_blank"}
-*   [zip](http://php.net/manual/en/book.zip.php){:target="_blank"}
-*	[json](http://php.net/manual/en/book.json.php){:target="_blank"}
-*	[iconv](http://php.net/manual/en/book.iconv.php){:target="_blank"}
+* [bc-math](http://php.net/manual/en/book.bc.php){:target="_blank"} ({{site.data.var.ee}} only for 2.2.0 - 2.2.3. {{site.data.var.ee}} and {{site.data.var.ce}} as of 2.2.4.)
```
[the remark this answers is not in the document]
Comment (dobooth) [reply guides/v2.2/install-gde/system-requirements-tech.md]: Done.

File: guides/v2.2/config-guide/elasticsearch/es-downgrade.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (dobooth) [reply guides/v2.2/config-guide/elasticsearch/es-downgrade.md]: I had it as "Downgrade" but the instructions are actually for changing versions: Commands are the same in either direction, so I changed it.

File: guides/v2.2/config-guide/elasticsearch/es-downgrade.md [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+
+If you need to run Magento 2.2.8 with Elasticsearch 2.x or 5.x, you must change the Elasticsearch [PHP client][] version.
+
+Both ES 2.x and 5.x are [End of Life][].
+Running ES 2.x is strongly discouraged.
+
+## Change the Elasticsearch Client version
```
[the remark this answers is not in the document]
Comment (dobooth) [reply guides/v2.2/config-guide/elasticsearch/es-downgrade.md]: Discussed above.

File: guides/v2.1/config-guide/elasticsearch/es-downgrade.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (dobooth) [reply guides/v2.1/config-guide/elasticsearch/es-downgrade.md]: Discussed above.

File: _includes/config/es-elasticsearch-magento.md [hunk 17]
```diff
@@ -1,102 +1,96 @@
-This section discusses the minimum settings you must choose to test Elasticsearch with Magento 2. For additional details about configuring Elasticsearch, see the [{{site.data.var.ee}} User Guide](http://docs.magento.com/m2/ee/user_guide/catalog/search-elasticsearch.html).
+This section discusses the minimum settings you must choose to test Elasticsearch with Magento 2.
+For additional details about configuring Elasticsearch, see the [{{site.data.var.ee}} User Guide](http://docs.magento.com/m2/ee/user_guide/catalog/search-elasticsearch.html).
+
+{: .bs-callout .bs-callout-warning}
+Magento 2.3.1 adds support for Elasticsearch 6.x and it is enabled by default.
+Magento still provides modules for Elasticsearch 2.x and 5.x but these must be enabled in order to use these versions.
+Elasticsearch 2.x is still available but strongly discouraged. Versions 2.x and 5.x are [End of Life](https://www.elastic.co/support/eol).
+If you must run Elasticsearch 2.x or 5.x with Magento 2.3.1, you must change the Elasticsearch client.
+Follow the instructions in [Change Elasticsearch Client]({{page.baseurl}}/config-guide/elasticsearch/es-downgrade.html).
```
[the remark this answers is not in the document]
Comment (dobooth) [reply _includes/config/es-elasticsearch-magento.md]: Keeping name for now.

File: _includes/config/es-elasticsearch-magento.md [hunk 18]
```diff
@@ -1,102 +1,96 @@
...
-</table>
+1. Log in to the Magento Admin as an administrator.
+1. Click **Stores** > Settings > **Configuration** > **Catalog** > **Catalog** > **Catalog Search**.
+1. From the **Search Engine** list, select the correct Elasticsearch version as the following figure shows.
+   (The **Elasticsearch 5.0+/6.0+** options are not available for Magento 2.1.)
+
+![]({{ site.baseurl }}/common/images/elastic_choose-in-admin-es6.png){:width="650px"}
+
+The following table discusses only the configuration options required to test the connection with Magento.
+Unless you changed Elasticsearch server settings, the defaults should work. Skip to the next step.
```
[the remark this answers is not in the document]
Comment (dobooth) [reply _includes/config/es-elasticsearch-magento.md]: They do not. It seems to work very well with defaults. This table is only if they are changing defaults.

File: _includes/config/es-elasticsearch-magento.md [hunk 19]
```diff
@@ -1,102 +1,96 @@
...
+Continue with:
+
+* [Configure Apache and Elasticsearch]({{ site.gdeurl21 }}config-guide/elasticsearch/es-config-apache.html)
+* [Configure nginx and Elasticsearch]({{ site.gdeurl21 }}config-guide/elasticsearch/es-config-nginx.html)
+
+or you will see:
+
+<img src="{{ site.baseurl }}/common/images/elastic_test-fail.png">
+
+If so, try the following:
```
[the remark this answers is not in the document]
Comment (dobooth) [reply _includes/config/es-elasticsearch-magento.md]: I don't want to put a heading in the middle of steps...

File: _includes/config/es-elasticsearch-magento.md [hunk 20]
```diff
@@ -1,102 +1,96 @@
...
+
+![]({{ site.baseurl }}/common/images/elastic_choose-in-admin-es6.png){:width="650px"}
+
+The following table discusses only the configuration options required to test the connection with Magento.
+Unless you changed Elasticsearch server settings, the defaults should work. Skip to the next step.
+
+|Option|Description|
+|--- |--- |
+|Elasticsearch Server Hostname|Enter the fully qualified hostname or IP address of the machine running Elasticsearch. {{site.data.var.ece}}: Get this value from your integration system.|
+|Elasticsearch Server Port|Enter the Elasticsearch web server proxy port. In our example, the port is 8080 but if you're using a secure proxy, it's typically 443. {{site.data.var.ece}}: Get this value from your integration system.|
```
[the remark this answers is not in the document]
Comment (dobooth) [reply _includes/config/es-elasticsearch-magento.md]: Bah! Some snuck through!

File: guides/v2.2/config-guide/elasticsearch/es-downgrade.md [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+composer require "elastic/elasticsearch:~5.1"
+```
+
+If you need to work with Elasticsearch 2.x, run the following command:
+
+```bash
+composer require "elastic/elasticsearch:~2.0"
+```
+
+To re-enable Elasticsearch 6.x, run the following command:
```
[the remark this answers is not in the document]
Comment (dobooth) [reply guides/v2.2/config-guide/elasticsearch/es-downgrade.md]: Added info block.

File: _includes/config/es-elasticsearch-magento.md [hunk 22]
```diff
@@ -1,102 +1,98 @@
...
+|**Elasticsearch** Server Hostname|Enter the fully qualified hostname or IP address of the machine running Elasticsearch. {{site.data.var.ece}}: Get this value from your integration system.|
+|**Elasticsearch Server Port**|Enter the Elasticsearch web server proxy port. In our example, the port is 8080 but if you are using a secure proxy, it is typically 443. {{site.data.var.ece}}: Get this value from your integration system.|
+|**Elasticsearch Index Prefix**|Enter the Elasticsearch index prefix. If you use a single Elasticsearch instance for more than one Magento installation (Staging and Production environments), you must specify a unique prefix for each installation. Otherwise, you can use the default prefix magento2.|
+|**Enable Elasticsearch HTTP Auth**|Click **Yes** only if you enabled authentication for your Elasticsearch server. If so, provide a username and password in the provided fields.|
+{:style="table-layout:auto;"}
+
+{:start="4"}
+1. Click **Test Connection**.
+
+You will see either:
```
Comment (lorikrell) [inline _includes/config/es-elasticsearch-magento.md]: Just a question, see either meaning the image and/or the sentence Magento successfully connected to the Elasticsearch server? If just the image, change to You will see a success message: or something like that.
If image or text, then add a blank line before Continue with:

Comment (dobooth) [reply _includes/config/es-elasticsearch-magento.md]: The either refers to either of the two image below: success or not.

**Pull request merged**